src/Interpreters/maskSensitiveInfoInQueryForLogging.cpp

```diff
@@ -0,0 +1,607 @@
+
+            if (password_specified)
+            {
+                /// Password is specified so we do our replacement:
+                /// remote('addresses_expr', db.table, 'user', 'password', ...) -> remote('addresses_expr', db.table, 'user', '[HIDDEN]', ...)
+                wipePasswordFromArgument(function, data, arg_num + 1);
+            }
+        }
+
+        static void wipePasswordFromEncryptionFunctionArguments(ASTFunction & function, Data & data)
+        {
+            if constexpr (check_only)
+            {
+                data.can_contain_password = true;
+                return;
+            }
+
+            /// We replace all arguments after 'mode' with '[HIDDEN]':
+            /// encrypt('mode', 'plaintext', 'key' [, iv, aad]) -> encrypt('mode', '[HIDDEN]')
+
+            size_t num_arguments;
+            if (!tryGetNumArguments(function, &num_arguments) || (num_arguments < 2))
+                return;
+
+            wipePasswordFromArgument(function, data, 1);
+
+            auto & arguments = assert_cast<ASTExpressionList &>(*function.arguments).children;
+            arguments.resize(2);
+        }
+
+        static void visitBackupQuery(ASTBackupQuery & query, Data & data)
+        {
+            if (query.backup_name)
+            {
+                if (auto * backup_engine = query.backup_name->as<ASTFunction>())
+                    wipePasswordFromBackupEngineArguments(*backup_engine, data);
+            }
+
+            if (query.base_backup_name)
+            {
+                if (auto * base_backup_engine = query.base_backup_name->as<ASTFunction>())
+                    wipePasswordFromBackupEngineArguments(*base_backup_engine, data);
+            }
+        }
+
+        static void wipePasswordFromBackupEngineArguments(ASTFunction & engine, Data & data)
+        {
+            if (engine.name == "S3")
+            {
+                /// BACKUP ... TO S3(url, [aws_access_key_id, aws_secret_access_key])
+                wipePasswordFromArgument(engine, data, 2);
+            }
+        }
+
+        static void wipePasswordFromArgument(ASTFunction & function, Data & data, size_t arg_idx)
+        {
+            if constexpr (check_only)
+            {
+                data.can_contain_password = true;
+                return;
+            }
+
+            if (!function.arguments)
+                return;
+
+            auto * expr_list = function.arguments->as<ASTExpressionList>();
+            if (!expr_list)
+                return;
+
+            auto & arguments = expr_list->children;
+            if (arg_idx >= arguments.size())
+                return;
+
+            arguments[arg_idx] = std::make_shared<ASTLiteral>("[HIDDEN]");
+        }
+
+        static bool tryGetNumArguments(const ASTFunction & function, size_t * num_arguments)
+        {
+            if (!function.arguments)
+                return false;
+
+            auto * expr_list = function.arguments->as<ASTExpressionList>();
```

Comment:
can we use assert_cast here ? also in tryGetStringFromArgument, tryGetEvaluatedConstStringFromArgument, visitDictionaryDef if ASTExpressionList is what we expect.

Reply:
as explained in above comment same applies here